### PR TITLE
fix: local directory submit

### LINKED ISSUE #6194: [BUG] mlflow run local filedirectory is broken in v1.27.0
### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

1.27.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS Monterey 12.3.1
- **Python version**: 3.7
- **yarn version, if running the dev UI**: N/A

### Describe the problem

The following command, which ran without issues in v1.26.1 fails in v1.27.0
```bash
mlflow run yolov5 --experiment-id <experiment-id> -b <backend> --backend-config yolov5/backend-spec.json
```

The log trace suggests that it assumes `yolov5` is a git repository link.

I checked the codebase: [`mlflow.projects.utils`](https://github.com/mlflow/mlflow/blob/v1.27.0/mlflow/projects/utils.py#L97)

```python
def _is_git_repo(path):
    """Returns True if passed-in path is a valid git repository"""
    import git

    try:
        git.Repo(path)
        return True
    except git.exc.InvalidGitRepositoryError:
        return False


def _is_local_uri(uri):
    """Returns True if passed-in URI should be interpreted as a path on the local filesystem."""
    if _GIT_URI_REGEX.match(uri):
        return False

    parsed_uri = urllib.parse.urlparse(uri)
    drive = pathlib.Path(uri).drive

    if drive != "" and drive.lower()[0] == parsed_uri.scheme:
        return not _is_git_repo(uri)
    elif parsed_uri.scheme in ("file", ""):
        return not _is_git_repo(parsed_uri.path)
    else:
        return False
```

Here are some quick checks:
```bash
>>> from pathlib import Path
>>> drive = Path("yolov5").drive
>>> drive
''
>>> import urllib
>>> urllib.parse.urlparse("yolov5")
ParseResult(scheme='', netloc='', path='yolov5', params='', query='', fragment='')
>>> import git
>>> git.Repo("yolov5")
<git.repo.base.Repo '/Users/cjagadeesan/workspace/python/yolov5/.git'>
```

It looks like both `parsed_uri.scheme in ("file", "")` and `_is_git_repo(parsed_uri.path)` are evaluated as `True` and therefore `False` is returned.   (`elif` condition in `_is_local_uri(uri)`)

Unsure about the purpose of checking this way: `Path(uri).exists()` would evaluate to `True` for local directories in both unix based filesytems and windows.  

**If `git init` has been run within the folder, this bug gets triggered** because `git.Repo(uri)` runs without issue  

```bash
mlflow run sample-mlflow -b local --experiment-name "sample-mlflow"

2022/07/05 08:53:39 INFO mlflow.utils.conda: Conda environment mlflow-da39a3ee5e6b4b0d3255bfef95601890afd80709 already exists.
2022/07/05 08:53:39 INFO mlflow.projects.utils: === Created directory /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmph5qx0jfd for downloading remote URIs passed to arguments of type 'path' ===
2022/07/05 08:53:39 INFO mlflow.projects.backend.local: === Running command 'source /opt/miniconda3/bin/../etc/profile.d/conda.sh && conda activate mlflow-da39a3ee5e6b4b0d3255bfef95601890afd80709 1>&2 && scripts/run.sh' in run with ID '5fb3639ad60f4abd8a96ccd7252d34b0' ===
Hey this ran!
2022/07/05 08:53:39 INFO mlflow.projects: === Run (ID '5fb3639ad60f4abd8a96ccd7252d34b0') succeeded ===

cd sample-mlflow
git init .

Initialized empty Git repository in /Users/cjagadeesan/workspace/python/sample-mlflow/.git/

cd ..
mlflow run sample-mlflow -b local --experiment-name "sample-mlflow"

2022/07/05 08:54:04 INFO mlflow.projects.utils: === Fetching project from sample-mlflow into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpj9o_m95d ===
Traceback (most recent call last):
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/bin/mlflow", line 8, in <module>
    sys.exit(cli())
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/cli.py", line 195, in run
    run_name=run_name,
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/__init__.py", line 343, in run
    run_name=run_name,
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/__init__.py", line 102, in _run
    experiment_id,
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/backend/local.py", line 64, in run
    work_dir = fetch_and_validate_project(project_uri, version, entry_point, params)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/utils.py", line 137, in fetch_and_validate_project
    work_dir = _fetch_project(uri=uri, version=version)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/utils.py", line 170, in _fetch_project
    _fetch_git_repo(parsed_uri, version, dst_dir)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/mlflow/projects/utils.py", line 220, in _fetch_git_repo
    output = g.execute(cmd)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/git/cmd.py", line 983, in execute
    raise GitCommandError(redacted_command, status, stderr_value, stdout_value)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git remote show origin
  stderr: 'fatal: 'sample-mlflow' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.'
```

### Tracking information

```python
MLflow version: 1.27.0
Tracking URI: databricks
Artifact URI: databricks
```

### Code to reproduce issue

```python
import mlflow
mlflow.projects.run(uri="nmt-char", backend="local", experiment_id="2315913529922077")
```

For your testing purposes, clone a repo locally. 
`git.Repo(uri)` should not throw an error.

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by ElefHead**
## Related Issues/PRs

Fixes #6194 

## What changes are proposed in this pull request?

Checks URIs not by regexes but by checking whether they exist on the system. Local git repos are checked in a more robust manner. 

1. Local filepaths are checked using `Path(ur).exists()`.
2. Local Git repos are checked by doing trying to see if a Repo class can be created and checking if the repo has branches.
    * This is because the Repo class can be created just by doing a git init in the file location, but without files actually being tracked and a branch being created, `git fetch` will fail.

## How is this patch tested?

Existing tests are sufficient, this code changes the way existing local folders are treated by the code.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
4. Click `Details` on the right to open the job page of CircleCI.
5. Click the `Artifacts` tab.
6. Click `docs/build/html/index.html`.
7. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/hyperparam/search_hyperopt.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 from hyperopt import fmin, hp, tpe, rand
 
 import mlflow.projects
-from mlflow import MlflowClient
+from mlflow.tracking import MlflowClient
```
Comment (harupy) [inline examples/hyperparam/search_hyperopt.py:20]: Can you revert unrelated changes?

File: mlflow/projects/utils.py [hunk 2]
```diff
@@ -83,31 +80,9 @@ def _is_file_uri(uri):
...
-    parsed_uri = urllib.parse.urlparse(uri)
-    drive = pathlib.Path(uri).drive
-
-    if drive != "" and drive.lower()[0] == parsed_uri.scheme:
-        return not _is_git_repo(uri)
-    elif parsed_uri.scheme in ("file", ""):
-        return not _is_git_repo(parsed_uri.path)
-    else:
-        return False
+    return pathlib.Path(uri).exists() and pathlib.Path(uri).is_dir()
```
Comment (harupy) [inline mlflow/projects/utils.py:85]: Does `pathlib.Path.exists` work properly when `uri` is a file URI?

```
>>> Path.cwd()
PosixPath('/home/haru/Desktop/repositories/mlflow')
>>> Path.cwd().as_uri()
'file:///home/haru/Desktop/repositories/mlflow'
>>> Path(Path.cwd().as_uri()).exists()
False
>>> Path(Path.cwd().as_uri(), "setup.py").exists()
False
```

File: examples/hyperparam/search_hyperopt.py (hunk 1 above)
Comment (ElefHead) [reply examples/hyperparam/search_hyperopt.py:20]: Oh yes absolutely

File: mlflow/projects/utils.py (hunk 2 above)
Comment (ElefHead) [reply mlflow/projects/utils.py:85]: I see what you mean, we could check that using the file uri regex, right? 
Let me make a small update.

Comment (ElefHead) [reply mlflow/projects/utils.py]: Done, I think this should work regardless of OS and should also work for file URIs.

File: mlflow/projects/utils.py [hunk 3]
```diff
@@ -96,18 +93,12 @@ def _is_git_repo(path):
...
-        return not _is_git_repo(uri)
-    elif parsed_uri.scheme in ("file", ""):
-        return not _is_git_repo(parsed_uri.path)
+    if _is_file_uri(str(uri)):
+        parsed_uri = urllib.parse.urlparse(str(uri))
+        resolved_uri = pathlib.Path(parsed_uri.netloc, parsed_uri.path).resolve()
     else:
-        return False
+        resolved_uri = pathlib.Path(uri).resolve()
+    return resolved_uri.exists()
```
Comment (ElefHead) [inline mlflow/projects/utils.py:101]: This should now take care of file URI types as well (and would be platform independent)

```bash
>>> from mlflow.projects.utils import _is_local_uri
>>> _is_local_uri(Path.cwd().as_uri())
True
>>> _is_local_uri(Path.cwd())
True
>>> _is_local_uri(Path(Path.cwd(), "setup.py").as_uri())
True
>>> _is_local_uri("https://github.com/mlflow/mlflow")
False
>>> _is_local_uri("git@github.com:mlflow/mlflow.git")
False
```

**Comment by harupy** [comment]
@ElefHead Thanks for the update! It looks like tests in the `python` and `python-skinny` checks failed. Can you fix them?

**Comment by ElefHead** [comment]
> @ElefHead Thanks for the update! It looks like tests in the `python` and `python-skinny` checks failed. Can you fix them?

@harupy Thanks for the review! It looks like the tests for _GIT_URI_REGEX are failing. I couldn't find anything using that regex outside of the `_is_local_uri(uri)` function, so I removed it. Should I put it back in, or can I remove those specific tests?

File: mlflow/projects/utils.py [hunk 4]
```diff
@@ -94,20 +91,21 @@ def _is_git_repo(path):
...
 
-def _is_local_uri(uri):
-    """Returns True if passed-in URI should be interpreted as a path on the local filesystem."""
-    if _GIT_URI_REGEX.match(uri):
-        return False
+def _parse_file_uri(uri: str) -> str:
+    """Converts file URIs to filesystem paths"""
+    parsed_uri = uri
+    if _is_file_uri(uri):
+        parsed_file_uri = urllib.parse.urlparse(str(uri))
```
Comment (harupy) [inline mlflow/projects/utils.py:98]:
```suggestion
        parsed_file_uri = urllib.parse.urlparse(uri)
```

`str` here looks redudant.

File: mlflow/projects/utils.py [hunk 5]
```diff
@@ -94,20 +91,21 @@ def _is_git_repo(path):
         return False
 
 
-def _is_local_uri(uri):
-    """Returns True if passed-in URI should be interpreted as a path on the local filesystem."""
-    if _GIT_URI_REGEX.match(uri):
-        return False
+def _parse_file_uri(uri: str) -> str:
+    """Converts file URIs to filesystem paths"""
+    parsed_uri = uri
```
Comment (harupy) [inline mlflow/projects/utils.py:96]: Do we need this assignmen? Can we do:

```python
def _parse_file_uri(uri: str) -> str:
    """Converts file URIs to filesystem paths"""
    if _is_file_uri(uri):
        parsed_file_uri = urllib.parse.urlparse(uri)
        return str(
            pathlib.Path(parsed_file_uri.netloc, parsed_file_uri.path, parsed_file_uri.fragment)
        )
    return uri
```

?

**Comment by harupy** [comment]
@ElefHead Thanks for the update! It looks like all the relevant unit tests passed :) Can you check the DCO check failure?

File: mlflow/projects/utils.py (hunk 5 above)
Comment (ElefHead) [reply mlflow/projects/utils.py:96]: yup! Let me apply it

File: mlflow/projects/utils.py [hunk 6]
```diff
@@ -94,20 +91,20 @@ def _is_git_repo(path):
...
-    parsed_uri = urllib.parse.urlparse(uri)
-    drive = pathlib.Path(uri).drive
 
-    if drive != "" and drive.lower()[0] == parsed_uri.scheme:
-        return not _is_git_repo(uri)
-    elif parsed_uri.scheme in ("file", ""):
-        return not _is_git_repo(parsed_uri.path)
-    else:
-        return False
+def _is_local_uri(uri) -> bool:
```
Comment (harupy) [inline mlflow/projects/utils.py:104]:
```suggestion
def _is_local_uri(uri: str) -> bool:
```

nit

**Comment by ElefHead** [comment]
Could you help me figure out the `python` failing tests? I'd like to understand the desired behavior and what's being missed there.

**Comment by harupy** [comment]
```
E           git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
E             cmdline: git remote show origin
E             stderr: 'fatal: 'tests/resources/example_project' does not appear to be a git repository
E           fatal: Could not read from remote repository.
E           
E           Please make sure you have the correct access rights
E           and the repository exists.'
```

It looks like `_fetch_git_repo` assumes the passed-in URI is a **remote** git URI (e.g. https://github.com/mlflow/mlflow.git), but we pass a local path.

**Comment by ElefHead** [comment]
I think I identified it, fingers crossed.

**Comment by ElefHead** [comment]
So, let's say localpath has git initialized and has branches - and a user is trying to run that project with version set as `some_branch_name`, what would be the behavior in the scenario?

**Comment by ElefHead** [comment]
@harupy have a feeling its now fixed. Fingers crossed.

1. Local zipfile
```bash
>>> local_zip_file = str(Path("/Users/cjagadeesan/Downloads/bootstrap-4.0.0.zip"))
>>> _fetch_project(local_zip_file)
2022/07/08 15:12:29 INFO mlflow.projects.utils: === Fetching project from /Users/cjagadeesan/Downloads/bootstrap-4.0.0.zip into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpvc14qdry ===
'/var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpvc14qdry'
```

2. local zipfile as file URI
```bash
>>> local_zip_file_uri = Path(local_zip_file).as_uri()
>>> local_zip_file_uri
'file:///Users/cjagadeesan/Downloads/bootstrap-4.0.0.zip'
>>> _fetch_project(local_zip_file_uri)
2022/07/08 15:13:29 INFO mlflow.projects.utils: === Fetching project from file:///Users/cjagadeesan/Downloads/bootstrap-4.0.0.zip into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpfmbkq7ty ===
'/var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpfmbkq7ty'
```

3. remote zip file
```bash
>>> _fetch_project("https://github.com/twbs/bootstrap/releases/download/v4.0.0/bootstrap-4.0.0-dist.zip")
2022/07/08 15:12:16 INFO mlflow.projects.utils: === Fetching project from https://github.com/twbs/bootstrap/releases/download/v4.0.0/bootstrap-4.0.0-dist.zip into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpllq_im1y ===
'/var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpllq_im1y'
```

3. local folder
```bash
>>> root = Path.cwd().parent
>>> local_folder = root / "test-yolov5"
>>> _fetch_project(str(local_folder))
'/Users/cjagadeesan/workspace/python/test-yolov5'
```

4. local git repo
```bash
>>> local_git_repo = root / "elefhead-yolov5"         
>>> local_git_repo.exists()
True
>>> _fetch_project(str(local_git_repo))
'/Users/cjagadeesan/workspace/python/elefhead-yolov5'
>>> _fetch_project(str(local_git_repo), "master") # with version 
'/Users/cjagadeesan/workspace/python/elefhead-yolov5'
```

5. local git repo as file URI
```bash
>>> _fetch_project(f'{str(local_git_repo.as_uri())}#utils')
'/Users/cjagadeesan/workspace/python/elefhead-yolov5/utils'
>>> _fetch_project(f'{str(local_git_repo.as_uri())}#utils', "master")
'/Users/cjagadeesan/workspace/python/elefhead-yolov5/utils'
>>> _fetch_project(f'{str(local_git_repo.as_uri())}#utils', None)
'/Users/cjagadeesan/workspace/python/elefhead-yolov5/utils'
```

6. Remote Git with version and folder
```bash
>>> _fetch_project("https://github.com/mlflow/mlflow-export-import#databricks_notebooks", 'issue-2_test-2-tracking-servers')
2022/07/08 15:32:22 INFO mlflow.projects.utils: === Fetching project from https://github.com/mlflow/mlflow-export-import#databricks_notebooks into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpb91tid8t ===
'/var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpb91tid8t/databricks_notebooks'
>>> _fetch_project("https://github.com/mlflow/mlflow-example", 'test-4838')
2022/07/08 15:29:21 INFO mlflow.projects.utils: === Fetching project from https://github.com/mlflow/mlflow-example into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpc7rinv98 ===
'/var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpc7rinv98'
```

7. Failure mode for remote pull
```bash
>>> _fetch_project("https://github.com/mlflow/mlflow-export-import#..", 'issue-2_test-2-tracking-servers')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/cjagadeesan/workspace/python/mlflow/mlflow/projects/utils.py", line 165, in _fetch_project
    parsed_uri, subdirectory = _parse_subdirectory(uri)
  File "/Users/cjagadeesan/workspace/python/mlflow/mlflow/projects/utils.py", line 58, in _parse_subdirectory
    raise ExecutionException("'.' is not allowed in project subdirectory paths.")
mlflow.exceptions.ExecutionException: '.' is not allowed in project subdirectory paths.
```

8. Failure mode for invalid version
```bash
>>> _fetch_project("https://github.com/mlflow/mlflow-example", 'test-branch')
2022/07/08 15:27:55 INFO mlflow.projects.utils: === Fetching project from https://github.com/mlflow/mlflow-example into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpms6_eeeg ===
Traceback (most recent call last):
  File "/Users/cjagadeesan/workspace/python/mlflow/mlflow/projects/utils.py", line 227, in _fetch_git_repo
    origin.fetch(refspec=version, depth=GIT_FETCH_DEPTH)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/git/remote.py", line 868, in fetch
    kill_after_timeout=kill_after_timeout)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/git/remote.py", line 732, in _get_fetch_info_from_stderr
    proc.wait(stderr=stderr_text)
  File "/Users/cjagadeesan/.pyenv/versions/3.7.10/lib/python3.7/site-packages/git/cmd.py", line 501, in wait
    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git fetch -v --depth=1 origin test-branch
  stderr: 'fatal: couldn't find remote ref test-branch'
```

9. Failure mode for fake subdir
```bash
>>> _fetch_project("https://github.com/mlflow/mlflow-export-import#fake", 'issue-2_test-2-tracking-servers')
2022/07/08 15:52:32 INFO mlflow.projects.utils: === Fetching project from https://github.com/mlflow/mlflow-export-import#fake into /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpjn382npk ===
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/cjagadeesan/workspace/python/mlflow/mlflow/projects/utils.py", line 193, in _fetch_project
    raise ExecutionException("Could not find subdirectory %s of %s" % (subdirectory, dst_dir))
mlflow.exceptions.ExecutionException: Could not find subdirectory fake of /var/folders/yv/k41cfcdj5l12xbyjqwv7tz_80000gr/T/tmpjn382npk
```

10. Failure mode for local dir with version
```bash
>>> _fetch_project("../../python", "master")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/cjagadeesan/workspace/python/mlflow/mlflow/projects/utils.py", line 184, in _fetch_project
    raise ExecutionException("Setting version is only supported for Git project URIs")
mlflow.exceptions.ExecutionException: Setting version is only supported for Git project URIs
```

**Comment by harupy** [comment]
> So, let's say localpath has git initialized and has branches - and a user is trying to run that project with version set as `some_branch_name`, what would be the behavior in the scenario?

I think we should checkout in this case.

**Comment by ElefHead** [comment]
> I think we should checkout in this case.

I think that's what I'm doing but being stumped by the tests here :( 
Still trying.

**Comment by ElefHead** [comment]
@harupy oh yay! All tests pass!

**Comment by harupy** [comment]
@ElefHead Awesome! I'll take a look at the update :)

**Comment by ElefHead** [comment]
> @ElefHead Awesome! I'm taking a look at the update :)

Please let me know if I need to do something else to help this get merged. Thanks for the guidance :)

File: mlflow/projects/utils.py [hunk 7]
```diff
@@ -150,22 +149,23 @@ def _fetch_project(uri, version=None):
...
         _logger.info("=== Fetching project from %s into %s ===", uri, dst_dir)
     if _is_zip_uri(parsed_uri):
-        if _is_file_uri(parsed_uri):
-            parsed_file_uri = urllib.parse.urlparse(urllib.parse.unquote(parsed_uri))
-            parsed_uri = os.path.join(parsed_file_uri.netloc, parsed_file_uri.path)
+        parsed_uri = _parse_file_uri(parsed_uri)
         _unzip_repo(
             zip_file=(parsed_uri if _is_local_uri(parsed_uri) else _fetch_zip_repo(parsed_uri)),
-            dst_dir=dst_dir,
+            dst_dir=_parse_file_uri(dst_dir),
```
Comment (harupy) [inline mlflow/projects/utils.py:160]:
```suggestion
            dst_dir=dst_dir,
```

Is this `_parse_file_uri` redundant?

Comment (ElefHead) [reply mlflow/projects/utils.py:160]: yes

**Review by harupy — APPROVED** [review]
LGTM, thanks for the contribution!

**Comment by ElefHead** [comment]
Thank you! When could this be merged?

**Comment by harupy** [comment]
Merged!
